### PR TITLE
POLIO-1333 vaccine stock management backend list view api

### PR DESCRIPTION
**Pull request by mathvdh**
APIs and test of this APIs for the list view of Vaccine Stock Management Page.
Based on #964 

Related JIRA tickets : POLIO-1333, POLIO-1332, POLIO-1148, POLIO-1149

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] My migrations file are included
- [x] Are there enough tests

## Changes

Add APIs for the list view of Vaccine Stock Management.
Make use of the models created in POLIO-1332 (See #964 )
Quite complex calculations in the API to calculate the `vials_received`, `vials_used`, `stock_of_usable_vials` ...

## How to test

See `plugins/polio/tests/test_vaccine_stock_management.py` for details 
But basically you need an `Account`, `DataSource`, `SourceVersion`, `Project` and a `OrgUnit` and its related `OrgUnitType` (COUNTRY).
Then a `Campaign` and a `Round`. (for the Same Country)
A `VaccineRequestForm` and related `VaccineArrivalReport`.
Then a `VaccineStock` object from the account, country, vaccine.
Then some `OutgoingStockMovement`, `DestructionReport` and `IncidentReport` to have enough data to calculate all the values.
(See in `setUpTestData` in the test).

### CODE REVIEW

File: plugins/polio/api/vaccines/stock_management.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        return sum(movement.usable_vials_used for movement in stock_movements)
+
+    def get_stock_of_usable_vials(self, obj):
+        return self.get_vials_received(obj) - self.get_vials_used(obj)
+
+    def get_leftover_ratio(self, obj):
+        # Assuming leftover_ratio is the percentage of vials_used out of vials_received
+        vials_received = self.get_vials_received(obj)
+        vials_used = self.get_vials_used(obj)
+        return (vials_used / vials_received * 100) if vials_received else 0
```
Comment (quang-le) [inline plugins/polio/api/vaccines/stock_management.py:59]: maybe round after 2 decimals

File: plugins/polio/api/vaccines/stock_management.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    def get_stock_of_unusable_vials(self, obj):
+        destruction_reports = DestructionReport.objects.filter(vaccine_stock=obj)
+        incident_reports = IncidentReport.objects.filter(vaccine_stock=obj)
+        stock_movements = OutgoingStockMovement.objects.filter(vaccine_stock=obj)
+        return (
+            sum(report.unusable_vials_destroyed for report in destruction_reports)
+            + sum(report.unusable_vials for report in incident_reports)
+            + sum(movement.unusable_vials for movement in stock_movements)
+        )
```
Comment (quang-le) [inline plugins/polio/api/vaccines/stock_management.py:69]: This is not correct: the vials destroyed don't count as unusable: they have been removed from the inventory.
Not sure about the different incidents:
- It seems that "Return"  adds to usable vials
- I don't know if the stolen ones are counted as unusable or removed from the inventory, like the destroyed ones
- Same with losses
- And I have no idea what "Physical inventory"  is supposed to mean

So I think there's some clarification to be done

Comment (mathvdh) [reply plugins/polio/api/vaccines/stock_management.py:69]: @quang-le yes agreed for the needed clarifications

**Review by kemar — APPROVED** [review]
LGTM AFAICT

The review is a bit hard because a lot of changes were made in `POLIO-1332`. Maybe the next time you could set it as the target of the PR:

merge

Also the serializer is performing a lot of SQL queries. It could become slow when used in "list" mode.

**Pull request merged**